[WebAssembly] Fold any/alltrue (setcc x, 0, eq/ne) to [not] any/alltrue x

badumbatish opened the pull request:
Fixes https://github.com/llvm/llvm-project/issues/50142, a miss of further vectorization, where we can only achieve zext (xor (any_true), -1). 

Now in test case simd-setcc-reductions, it's converted to all_true.

Also fixes https://github.com/llvm/llvm-project/issues/145177, which is

all_true (setcc x, 0, eq) -> not any_true
any_true (setcc x, 0, ne) -> any_true
all_true (setcc x, 0, ne) -> all_true

dschuff commented:
@tlively is probably good (unless he wants to delegate that off), also @sparker-arm has been working on wasm SIMD recently.

lukel97 commented on llvm/test/CodeGen/WebAssembly/simd-setcc-reductions.ll:
```diff
@@ -0,0 +1,22 @@
+; RUN: llc < %s -asm-verbose=false -verify-machineinstrs -disable-wasm-fallthrough-return-opt -wasm-disable-explicit-locals -wasm-keep-registers -mattr=+simd128 | FileCheck %s
```
You probably want to use ./utils/update_llc_test_checks.py to generate these, since they're much easier to update if any of the codegen ever changes, I think simd-arith.ll has an example of what it looks like

lukel97 commented on llvm/test/CodeGen/WebAssembly/simd-setcc-reductions.ll:
```diff
@@ -0,0 +1,22 @@
+;CHECK-NEXT:        return  $pop1
+;CHECK-NEXT:        end_function
+; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(argmem: read)
+define hidden range(i32 0, 2) i32 @all_true(ptr noundef readonly captures(none) %a) local_unnamed_addr #0 {
+entry:
+  %0 = load <16 x i8>, ptr %a, align 1
+  %.fr = freeze <16 x i8> %0
+  %1 = icmp eq <16 x i8> %.fr, zeroinitializer
```
The load and freeze shouldn't be relative for this test so you can probably just pass the vector in:

```suggestion
define i32 @all_true(<16 x i8>) {
  %1 = icmp eq <16 x i8> %v, zeroinitializer
```

lukel97 commented on llvm/test/CodeGen/WebAssembly/simd-setcc-reductions.ll:
```diff
@@ -0,0 +1,22 @@
+  %0 = load <16 x i8>, ptr %a, align 1
+  %.fr = freeze <16 x i8> %0
+  %1 = icmp eq <16 x i8> %.fr, zeroinitializer
+  %2 = bitcast <16 x i1> %1 to i16
```
It would probably also be good to add tests for the other vector shapes, v8i16/v4i32,v2i64

lukel97 reviewed (commented):
Can you [precommit](https://llvm.org/docs/InstCombineContributorGuide.html#precommit-tests) the test in this PR? I.e. rebase this so there's one commit adding the test without the changes, and another adding the changes to WebAssemblyISelLowering.cpp and the changes to the test case. That way you can see the diff in the PR

badumbatish committed:
Precommit missed optimization test for #50142

lukel97 commented on llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:
```diff
@@ -3248,6 +3248,37 @@ static SDValue performSETCCCombine(SDNode *N,
+  //  N           LHS     LhsL        LhsLL    LhsLR   InnerCond RHS Cond
+  // setcc (iN (bitcast (setcc vNi1 (vNiY X), <vNiY 0>, eq)),     0, eq
+  // => all_true (vNi1 X)
+  if (DCI.isBeforeLegalize() && VT.isScalarInteger() && (Cond == ISD::SETEQ) &&
```
Looking at this combine its quite similar to the one below, I wonder if it could be generalized?

i.e. we could rely on the combine below to do the first half, and then we would add a combine for

`wasm_anytrue (sext/zext (setcc x, 0, eq)) -> not (wasm_alltrue x)`

and

`wasm_anytrue (sext/zext (setcc x, 0, ne)) -> wasm_alltrue x`

I think you'd have then do the combine on `ISD::INTRINSIC_WO_CHAIN`. But this is just an idea, it might turn out to be harder :)

sparker-arm replied:
I found it somewhat painful to read the existing code, for which I have now suggested a refactor in #144875 144875.

badumbatish replied:
Ah I see, thank you for the refactor! i'll base this pr on top of yours now

lukel97 replied:
Since @sparker-arm's patch is NFC and I don't think you need to change the existing combine, I don't think you need to stack/rebase your patch on top of it.
We don't really have a good system for stacking patches in LLVM anyway :)

badumbatish committed:
Fix issue 50142 by adding AnyTrueCombine

This introduces the fold (any_true (setcc <X> 0, eq)) to (not
(all_true)), allowing potential extra fold of (not (not ...))

Introduces test simd-setcc-reductions and readjusts simd-vecreduce-bool

badumbatish commented:
i updated the code, in the case of 4xi16, the fold doesn't work since i was targeting setcc
```
define i32 @all_true_4_i16(<4 x i16> %v) {
; CHECK-LABEL: all_true_4_i16:
; CHECK:         .functype all_true_4_i16 (v128) -> (i32)
; CHECK-NEXT:  # %bb.0:
; CHECK-NEXT:    v128.const $push0=, 0, 0, 0, 0, 0, 0, 0, 0
; CHECK-NEXT:    i16x8.eq $push1=, $0, $pop0
; CHECK-NEXT:    i32x4.extend_low_i16x8_s $push2=, $pop1
; CHECK-NEXT:    v128.any_true $push3=, $pop2
; CHECK-NEXT:    i32.const $push4=, -1
; CHECK-NEXT:    i32.xor $push5=, $pop3, $pop4
; CHECK-NEXT:    i32.const $push6=, 1
; CHECK-NEXT:    i32.and $push7=, $pop5, $pop6
; CHECK-NEXT:    return $pop7
  %1 = icmp eq <4 x i16> %v, zeroinitializer
  %2 = bitcast <4 x i1> %1 to i4
  %3 = icmp eq i4 %2, 0
  %conv3 = zext i1 %3 to i32
  ret i32 %conv3
}
```

lukel97 commented on llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:
```diff
@@ -3400,8 +3436,11 @@ WebAssemblyTargetLowering::PerformDAGCombine(SDNode *N,
     return performTruncateCombine(N, DCI);
-  case ISD::INTRINSIC_WO_CHAIN:
+  case ISD::INTRINSIC_WO_CHAIN: {
+    if (auto AnyTrueCombine = performAnyTrueCombine(N, DCI.DAG))
```
Nit, auto is mainly only used when the type already appears in a `cast<Foo>` or the type is complex like an std::pair
```suggestion
    if (SDValue AnyTrueCombine = performAnyTrueCombine(N, DCI.DAG))
```

lukel97 commented on llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:
```diff
@@ -3239,6 +3239,42 @@ static SDValue performBitcastCombine(SDNode *N,
+  if (N->getConstantOperandVal(0) != Intrinsic::wasm_anytrue)
+    return SDValue();
+
+  SDValue SetCC = N->getOperand(1);
+  if (SetCC.getOpcode() != ISD::SETCC)
+    return SDValue();
+
+  SDValue LHS = SetCC->getOperand(0);
+  SDValue RHS = SetCC->getOperand(1);
+  ISD::CondCode Cond = cast<CondCodeSDNode>(SetCC->getOperand(2))->get();
+  EVT LT = LHS.getValueType();
+  unsigned NumElts = LT.getVectorNumElements();
+  if (NumElts != 2 && NumElts != 4 && NumElts != 8 && NumElts != 16)
+    return SDValue();
+
+  EVT Width = MVT::getIntegerVT(128 / NumElts);
+
+  if (!isNullOrNullSplat(RHS) || Cond != ISD::SETEQ)
+    return SDValue();
```
This would be a good place to use the pattern matching framework for SelectionDAG:

```suggestion
  SDValue Vec;
  if (!sd_match(
          N->getOperand(1), m_c_SetCC(m_Value(Vec), m_Zero(), m_SpecificCondCode(ISD::SETEQ)))
    return false;
```

lukel97 commented on llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:
```diff
@@ -3239,6 +3239,42 @@ static SDValue performBitcastCombine(SDNode *N,
+  ISD::CondCode Cond = cast<CondCodeSDNode>(SetCC->getOperand(2))->get();
+  EVT LT = LHS.getValueType();
+  unsigned NumElts = LT.getVectorNumElements();
+  if (NumElts != 2 && NumElts != 4 && NumElts != 8 && NumElts != 16)
+    return SDValue();
```
If we're matching on an any_true intrinsic does the number of elements not always have to be 2/4/8/16 to begin with? Can we get away with removing this check?

lukel97 commented on llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:
```diff
@@ -3239,6 +3239,42 @@ static SDValue performBitcastCombine(SDNode *N,
+      DAG.getNode(
+          ISD::INTRINSIC_WO_CHAIN, DL, MVT::i32,
+          {DAG.getConstant(Intrinsic::wasm_alltrue, DL, MVT::i32),
+           DAG.getSExtOrTrunc(LHS, DL, LT.changeVectorElementType(Width))}),
```
I don't think we can truncate the underlying vector, e.g. if we had `any_true (setcc (v4i64 x), 0, eq)` and we were to fold that  `not (all_true (trunc (v4i64 x)))` then we would return the wrong result if any of the bits 32 and above were set.

I think we need to bail if `LT.getScalarSizeInBits() > Width`, can you also add a test for this case?

badumbatish replied:
hmm ok, I removed the truncation and added the condition to return early as well.

Is this satisfied?

lukel97 replied:
Thanks, I think so. Just still need a test case for v4i64!

lukel97 reviewed (commented):
Thanks, this new approach looks good!

badumbatish committed:
Use SDPatternMatching and remove truncate...

Use SDPatternMatching and remove truncation. Also added 4xi64 case to
reflect that.

lukel97 commented on llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:
```diff
@@ -35,9 +36,10 @@
 #include "llvm/Target/TargetOptions.h"
 using namespace llvm;
-
+using namespace llvm::SDPatternMatch;
```
You should move this into performAnyTrueCombine so it's not in the global scope

lukel97 commented on llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:
```diff
@@ -35,9 +36,10 @@
 #include "llvm/Support/ErrorHandling.h"
 #include "llvm/Support/KnownBits.h"
 #include "llvm/Support/MathExtras.h"
+#include "llvm/Support/raw_ostream.h"
```
Looks leftover from debugging? :)

lukel97 commented on llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:
```diff
@@ -3239,6 +3241,37 @@ static SDValue performBitcastCombine(SDNode *N,
+  if (LT.getScalarSizeInBits() > 128 / NumElts)
+    return SDValue();
+
+  EVT Width = MVT::getIntegerVT(128 / NumElts);
```
This doesn't look like it's used?

lukel97 commented on llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:
```diff
@@ -37,7 +38,6 @@
 #include "llvm/Support/MathExtras.h"
 #include "llvm/Target/TargetOptions.h"
 using namespace llvm;
-
```
Stray whitespace change

lukel97 reviewed (approved):
LGTM, thanks! Although we should also wait for a WebAssembly maintainer to approve this too (this isn't my neck of the woods in LLVM!)

Can you also update your PR title to not reference the issue, but instead something like "[WebAssembly] Combine any_true (setcc x, 0, eq) to not all_true"? 

We should also follow up on this to also handle:

- `all_true (setcc x, 0, eq) -> not any_true`
- `any_true (setcc x, 0, ne) -> any_true`
- `all_true (setcc x, 0, ne) -> all_true`

badumbatish commented:
really sorry about the extra ping @tlively , can I please get the PR either approved or reviewed? 

I think I can reuse some of the code for issue #145177. I just want to make sure the code structure is good before I base the new PR ontop of this PR

badumbatish committed:
Precommit test to add 3 more any/all true patterns

badumbatish committed:
[WebAssembly] Add 3 more optimization for any/all

all_true (setcc x, 0, eq) -> not any_true
any_true (setcc x, 0, ne) -> any_true
all_true (setcc x, 0, ne) -> all_true

badumbatish commented:
I've refactored, and added support and test case for three more patterns
```
all_true (setcc x, 0, eq) -> not any_true
any_true (setcc x, 0, ne) -> any_true
all_true (setcc x, 0, ne) -> all_true
```

lukel97 commented on llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:
```diff
@@ -3239,6 +3240,55 @@ static SDValue performBitcastCombine(SDNode *N,
+        DAG.getNode(ISD::INTRINSIC_WO_CHAIN, DL, MVT::i32,
+                    {DAG.getConstant(InPost, DL, MVT::i32), LHS}),
+        DL, MVT::i1);
+    if (ShouldInvert)
```
I think you can remove `ShouldInvert` if you just directly check the CondCode, 
```suggestion
    if (SetType == ISD::SETEQ)
```

lukel97 commented on llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:
```diff
@@ -3239,6 +3240,55 @@ static SDValue performBitcastCombine(SDNode *N,
+  assert(N->getOpcode() == ISD::INTRINSIC_WO_CHAIN);
+  using namespace llvm::SDPatternMatch;
+  SDLoc DL(N);
+  static auto SimdCombiner =
```
Nit, I don't think static lambdas are that common in LLVM. Also maybe a better name for this is like `CombineSetCC`?

lukel97 commented on llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:
```diff
@@ -3239,6 +3240,55 @@ static SDValue performBitcastCombine(SDNode *N,
+  using namespace llvm::SDPatternMatch;
+  SDLoc DL(N);
+  static auto SimdCombiner =
+      [&](Intrinsic::WASMIntrinsics InPre, ISD::CondCode SetType,
```
It's better to explicitly capture what you need, e.g.

```suggestion
      [&N, &DAG](Intrinsic::WASMIntrinsics InPre, ISD::CondCode SetType,
```

lukel97 commented on llvm/test/CodeGen/WebAssembly/simd-setcc-reductions.ll:
```diff
@@ -0,0 +1,136 @@
+
+; setcc (iN (bitcast (set_cc (vNi1 X), 0, eq)), -1, ne
+;   => not all_true (set_cc (X), 0, eq)
+;   => not all_true (set_cc (X), 0, eq)
```
Looks like a duplicate comment line?

```suggestion
```

lukel97 reviewed (commented):
Thanks for handling the new test cases! You should add "Fixes #50142, fixes #145177" to your PR description so it will automatically close those issues when this gets merged

lukel97 commented on llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:
```diff
@@ -3239,6 +3240,55 @@ static SDValue performBitcastCombine(SDNode *N,
+  using namespace llvm::SDPatternMatch;
+  SDLoc DL(N);
+  auto CombineSetCC =
+      [&N, &DAG, &DL](Intrinsic::WASMIntrinsics InPre, ISD::CondCode SetType,
+                      Intrinsic::WASMIntrinsics InPost) -> SDValue {
```
Nit, can the return type be inferred?
```suggestion
      [&N, &DAG, &DL](Intrinsic::WASMIntrinsics InPre, ISD::CondCode SetType,
                      Intrinsic::WASMIntrinsics InPost) {
```

lukel97 commented on llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:
```diff
@@ -3239,6 +3240,55 @@ static SDValue performBitcastCombine(SDNode *N,
+                                              m_SpecificCondCode(SetType))))
+      return SDValue();
+
+    EVT LT = LHS.getValueType();
+    unsigned NumElts = LT.getVectorNumElements();
+    if (LT.getScalarSizeInBits() > 128 / NumElts)
+      return SDValue();
```
Nit, I think this can be pulled out of the lambda and be put at the start of performAnyAllCombine

badumbatish replied:
will perform a precheck before the lambda

lukel97 commented on llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:
```diff
@@ -3239,6 +3240,55 @@ static SDValue performBitcastCombine(SDNode *N,
+  // all_true (setcc <X>, 0, ne) => (all_true X)
+  assert(N->getOpcode() == ISD::INTRINSIC_WO_CHAIN);
+  using namespace llvm::SDPatternMatch;
+  SDLoc DL(N);
```
Nit, this can be sunk into CombineSetCC to prevent the capture

lukel97 reviewed (approved):
LGTM with the new changes, thanks! Gentle ping to the other WebAssembly maintainers for another review

Commit:
[WebAssembly] Fix nit in PR 144741

dschuff commented on clang-tools-extra/test/clang-doc/json/class.cpp:
```diff
@@ -23,6 +23,9 @@ struct MyClass {
   typedef int MyTypedef;
```
This file looks like a merge artifact?

dschuff reviewed (approved):
Otherwise LGTM

Commit:
Merge branch 'main' of https://github.com/llvm/llvm-project into issue50142

dschuff commented:
It looks like this one is ready to go too. is the commit message up-to-date?

dschuff commented:
> just changed, tysm for merging

No problem, well done!

I'm just going to wait for the CI tests to finish, then I'll merge.

lukel97 commented:
Thanks for the review and the commit @dschuff, it's much appreciated!

lukel97 commented:
> ruh roh
> 
> 
> 
> i digged into the build log and couldn't find any failures, just that its timing out on task 84350???
> 
> ```
> 
> PASS: lit :: shtest-shell.py (84348 of 84350)
> 
> PASS: lit :: shtest-define.py (84349 of 84350)
> 
> ....
> 
> command timed out: 1200 seconds without output running [b'ninja', b'check-all'], attempting to kill
> 
> program finished with exit code 1
> 
> ```
> 
> 
> 
> how can i make sure this is not some kind of performance regression on my end?

I wouldn't worry about that buildbot failure, it's likely just a flakey test/bot. There's quite a lot of false alarms as you'll soon come to see :)
